options/options_helper.cc

```diff
@@ -1194,6 +1194,277 @@ Status GetTableFactoryFromMap(
   return Status::OK();
 }
 
+ColumnFamilyOptions OptionsHelper::dummy_cf_options;
+CompactionOptionsFIFO OptionsHelper::dummy_comp_options;
+
+// offset_of is used to get the offset of a class data member
+// ex: offset_of(&ColumnFamilyOptions::num_levels)
+// This call will return the offset of num_levels in ColumnFamilyOptions class
+//
+// This is the same as offsetof() but allow us to work with non standard-layout
+// classes and structures
+// refs:
+// http://en.cppreference.com/w/cpp/concept/StandardLayoutType
+// https://gist.github.com/graphitemaster/494f21190bb2c63c5516
+template <typename T1>
+int offset_of(T1 ColumnFamilyOptions::*member) {
+  return int(size_t(&(OptionsHelper::dummy_cf_options.*member)) -
+             size_t(&OptionsHelper::dummy_cf_options));
+}
+template <typename T1>
+int offset_of(T1 AdvancedColumnFamilyOptions::*member) {
+  return int(size_t(&(OptionsHelper::dummy_cf_options.*member)) -
+             size_t(&OptionsHelper::dummy_cf_options));
+}
+template <typename T1>
+int offset_of(T1 CompactionOptionsFIFO::*member) {
```

Comment:
Why do we need three offset_of overloads now? Doesn't the old definition of offset_of, with two template params, work here?

Reply:
>  It also moves the static variable out of the offset_of since the template function could also lead to multiple instantiation perhaps due to inlining.
With the general template we would not know which static variable to use, unless you know a c++ trick for that?

Reply:
syntax-wise this will work. but I didn't check if the offset returned is correct.

```
template <typename T>
struct Dummy {
  static const T dummy;
};

template <typename T>
const T Dummy<T>::dummy{};

template <typename T1, typename T2>
int my_offset_of(T1 T2::*member) {
  return int(size_t(&(Dummy<T2>::dummy.*member)) - size_t(&Dummy<T2>::dummy));
}
```

Reply:
@maysamyabandeh : Thanks for explaining; I missed reading that line in the summary. Your solution is fine with me, based on this new information.